### PR TITLE
kubeadm: Adds default etcd certificate locations

### LINKED ISSUE #1115: kubeadm init api-server ignores etcd-cafile etcd-certfile etcd-keyfile during manifest write.
BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`): kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:14:41Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): ubuntu
- **Kernel** (e.g. `uname -a`): Linux 4.4.0-1066-aws #76-Ubuntu SMP Thu Aug 16 16:21:21 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- **Others**:

## What happened?
Kubeadm init does not provide CA / key / cert for etcd servers when an external server is used, which then causes the apiserver not to start up as it cannot connect to the external etcd servers.

## What you expected to happen?
During kubeadm init the manifests should be written with the etcd ca / cert / key information as switches.

## How to reproduce it (as minimally and precisely as possible)?
Create kubeadm-config.conf file with following data:
```
apiVersion: kubeadm.k8s.io/v1alpha2
kind: MasterConfiguration
kubernetesVersion: v1.11.0
certificatesDir: /etc/kubernetes/pki
api:
  controlPlaneEndpoint: "load-balancer-endpoint:443"
networking:
  podSubnet: 192.168.0.0/16
etcd:
  external:
    endpoints:
    - https://external-etcd-1.server.com:2379
    - https://external-etcd-2.server.com:2379
    - https://external-etcd-3.server.com:2379
bootstrapTokens:
- groups:
  - system:bootstrappers:kubeadm:default-node-token
  token: 1s1bvu.u1wll2dl2eo701ul
```

now run:
```kubeadm init --ignore-preflight-errors=all --config kubeadm-config.conf```

Content of /etc/kubernetes/manifests/kube-apiserver.yaml is now:
```
.... redacted ....
spec:
  containers:
  - command:
    - kube-apiserver
    - --authorization-mode=Node,RBAC
    - --advertise-address=REDACTED
    - --allow-privileged=true
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --disable-admission-plugins=PersistentVolumeLabel
    - --enable-admission-plugins=NodeRestriction
    - --enable-bootstrap-token-auth=true
    - --etcd-servers=https://external-etcd-1.server.com:2379,https://external-etcd-1.server.com:2379,https://external-etcd-1.server.com:2379
    - --insecure-port=0
    - --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt
    - --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key
    - --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
    - --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt
    - --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key
    - --requestheader-allowed-names=front-proxy-client
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --requestheader-extra-headers-prefix=X-Remote-Extra-
    - --requestheader-group-headers=X-Remote-Group
    - --requestheader-username-headers=X-Remote-User
    - --secure-port=6443
    - --service-account-key-file=/etc/kubernetes/pki/sa.pub
    - --service-cluster-ip-range=10.96.0.0/12
    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
    env:
```
Expecting:
```
    - --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt
    - --etcd-certfile=/etc/kubernetes/pki/etcd/server.crt
    - --etcd-keyfile=/etc/kubernetes/pki/etcd/server.key
``` 
To be part of the flags in command setup.

## Anything else we need to know?
Notice there is no config variables for setting apiserver etcd ca servers so the assumption is that it will use the certs that are pre-set under /etc/kubernetes/pki.  Kubelet and other services do however use the existing certs correctly.

### PR DESCRIPTION
**Pull request by chuckha**
kubeadm will use default etcd certificate locations if no
etcd certificates locations are supplied.

Removes a test that supports non-TLS etcd.

Signed-off-by: Chuck Ha <ha.chuck@gmail.com>

**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR sets etcd certificate locations to expected defaults so they do not need to be provided by the user.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#1115

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews 
@Nitecon

### CODE REVIEW

**Comment by chuckha** [comment]
/hold 

Turns out we do support non-TLS with external etcd.

**Review by neolit123 — COMMENTED** [review]
so having the default locations SGTM and that should be a fix for the linked issue, i assume.
not sure how to handle the non-TLS part though.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -153,14 +152,22 @@ func getAPIServerCommand(cfg *kubeadmapi.InitConfiguration) []string {
...
+		defaultArguments["etcd-cafile"] = filepath.Join(cfg.CertificatesDir, kubeadmconstants.EtcdCACertName)
+		defaultArguments["etcd-certfile"] = filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerEtcdClientCertName)
+		defaultArguments["etcd-keyfile"] = filepath.Join(cfg.CertificatesDir, kubeadmconstants.APIServerEtcdClientKeyName)
+
+		// If they are defined in the configuration, use that value instead.
 		if cfg.Etcd.External.CAFile != "" {
 			defaultArguments["etcd-cafile"] = cfg.Etcd.External.CAFile
 		}
-		if cfg.Etcd.External.CertFile != "" && cfg.Etcd.External.KeyFile != "" {
+		if cfg.Etcd.External.CertFile != "" {
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: IMO external cert and key should appear in pair. I'd be good to assume they are configured in the same time. WDYT?

**Comment by chuckha** [comment]
/hold cancel

Updated PR to cover external non-TLS etcd cases.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (chuckha) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: The logic behind this change was that if a user sets one but not the other they likely want that one overridden and the default for the other one. 

Before, if a user set one and not the other, neither would have a value (and no warning/messaging) which seems a bit confusing. Perhaps we need a warning if only one is set? Something like "Only Certfile or Keyfile was set. Both must be set or both must be empty. Assuming defaults." ?

**Comment by chuckha** [comment]
@neolit123 non-tls support is returned in this PR

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (xiangpengzhao) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: >The logic behind this change was that if a user sets one but not the other they likely want that one overridden and the default for the other one.

SGTM with leaving your change as-is :)

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
/hold
@chuckha @neolit123 IMO settings etcd certs defaults in the config and then ignoring them when generating API server command in case of http is weird.

Let me try to explain my considerations about this issue/pr.

In case of external etcd kubeadm does not have any kind of control on etcd certs naming and location (the user can choose any name for certs and save them in any location). So settings default values IMO doesn't make sense.

If we want to facilitate users that choose to use a well known naming for external CA certificates and choose to place certificates in a well known location, we can implement a dynamic default instead.

E.g. if external etcd and HTTPS and certificates are not set but existing in a well known location/with a well known name, then pick it up

A lot of if, but the use case is based on a multiple assumptions. This should be well documented somewhere (godoc for kubeadm config?)
Wdyt?

**Comment by neolit123** [comment]
@fabriziopandini 
i'm OK with documenting that the user has to set them manually, like a instructed the user here:
https://github.com/kubernetes/kubeadm/issues/1115#issuecomment-421145459

no strong opinions on this.

**Comment by chuckha** [comment]
@fabriziopandini I agree that setting and ignoring in the case of http is weird. I'm totally fine not merging this, I think your points make a lot of sense.

@neolit123 I see you have mentioned documenting this but I do not understand what needs to be documented as this documentation already exists here: https://kubernetes.io/docs/setup/independent/high-availability/#set-up-the-first-control-plane-node. Is this documentation lacking in some way that could be improved?

**Comment by fabriziopandini** [comment]
@chuckha If we are closing this PR I'm fine with the documentation as it is.

Instead, if we are implementing the dynamic default we should document somewhere all the assumption behind

**Comment by chuckha** [comment]
@fabriziopandini @neolit123 after thinking this one through I agree with you that what is here is weird. I think adding a godoc comment as well as making it more explicit in the HA docs will be good to do.

**Pull request closed without merging**